dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCResponseHeadersTest.java

```diff
@@ -0,0 +1,1490 @@
+import org.apache.hc.client5.http.classic.methods.HttpPost;
+import org.apache.hc.client5.http.entity.UrlEncodedFormEntity;
+import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
+import org.apache.hc.client5.http.impl.classic.CloseableHttpResponse;
+import org.apache.hc.client5.http.impl.classic.HttpClientBuilder;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.NameValuePair;
+import org.apache.hc.core5.http.io.entity.EntityUtils;
+import org.apache.hc.core5.http.message.BasicNameValuePair;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.Collections;
+import java.util.List;
+
+import com.ibm.websphere.simplicity.ShrinkHelper;
+import com.ibm.websphere.simplicity.config.HttpEndpoint;
+import com.ibm.websphere.simplicity.config.ServerConfiguration;
+import com.ibm.websphere.simplicity.log.Log;
+
+import componenttest.annotation.Server;
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.utils.ServerFileUtils;
+
+/**
+ * A set of tests to verify the <header> configurations and the corresponding response headers
+ */
+@RunWith(FATRunner.class)
+public class WCResponseHeadersTest {
+
+    private static final Class<?> ME = WCResponseHeadersTest.class;
+    private static final String APP_NAME = "ResponseHeadersTest";
+    private static final String APP_VERIFICATION_HEADER = "appVerificationHeader";
+    //App used to simulate responses during login process
+    private static final String APP_NAME_SECURE_APP = "SameSiteSecurityTest";
+
+    @Server("servlet40_headers")
+    public static LibertyServer server;
+
+    private static final ServerFileUtils serverFileUtils = new ServerFileUtils();
+
+   
+
+    // Tests can use this to indicate they don't make any config updates from the original configuration
+    private static boolean restoreSavedConfig = true; 
+
+    private static ServerConfiguration savedConfig; 
+
+    @Before
+    public void setUpBeforeEachTest() throws Exception {
+    
+        Exception failure = null;
+        String consoleLogFileName = WCResponseHeadersTest.class.getSimpleName() + ".log";
+        
+        if(!server.isStarted()){
+            server.updateServerConfiguration(savedConfig);
+            server.startServer(consoleLogFileName);
+            Log.info(ME, "setUpBeforeTest", "server started, log file is " + consoleLogFileName);
+        } else if(restoreSavedConfig){
+            try{
+                server.stopServer("CWWKT0042W", "CWWKT0043W", "CWWKT0044W");
+            } catch (Exception e){
+                failure = e;
+            }
+            server.updateServerConfiguration(savedConfig);
+            server.startServer(consoleLogFileName, true);
+            Log.info(getClass(), "setUpBeforeTest", "server restarted, log file is " +consoleLogFileName);
```

Comment:
In these start server calls do we need to ensure we're waiting for the config updates to complete? Something like:  `server.waitForConfigUpdateInLogUsingMark(Collections.singleton(appName), false/true, "CWWKT0016I:.*myAppName.*");`

Reply:
I don't believe this would work since it is not updating the config on a running server, rather starting one up that is shutdown. The application is verified to be started through the `startServer` API, which does a `startServerAndValidate` with `validateApps` set to true by default